tests/components/generic_thermostat/test_climate.py

```diff
@@ -1347,6 +1348,72 @@ async def test_restore_will_turn_off_(hass):
     assert hass.states.get(heater_switch).state == STATE_ON
 
 
+async def test_restore_will_turn_off_when_loaded_second(hass):
+    """Ensure that restored state is coherent with real situation.
+
+    Switch is not available until after component is loaded
+    """
+    heater_switch = "input_boolean.test"
+    mock_restore_cache(
+        hass,
+        (
+            State(
+                "climate.test_thermostat",
+                HVAC_MODE_HEAT,
+                {ATTR_TEMPERATURE: "18", ATTR_PRESET_MODE: PRESET_NONE},
+            ),
+            State(heater_switch, STATE_ON, {}),
+        ),
+    )
+
+    hass.state = CoreState.starting
+
+    await hass.async_block_till_done()
+    assert hass.states.get(heater_switch) is None
+
+    _setup_sensor(hass, 16)
+
+    await async_setup_component(
+        hass,
+        DOMAIN,
+        {
+            "climate": {
+                "platform": "generic_thermostat",
+                "name": "test_thermostat",
+                "heater": heater_switch,
+                "target_sensor": ENT_SENSOR,
+                "target_temp": 20,
+                "initial_hvac_mode": HVAC_MODE_OFF,
+            }
+        },
+    )
+    await hass.async_block_till_done()
+    state = hass.states.get("climate.test_thermostat")
+    assert state.attributes[ATTR_TEMPERATURE] == 20
+    assert state.state == HVAC_MODE_OFF
+
+    calls = []
+
+    @callback
+    def log_call(call):
+        """Log service calls."""
+        calls.append(call)
+
+    hass.services.async_register(ha.DOMAIN, SERVICE_TURN_ON, log_call)
```

Comment:
We have a mock service helper:
https://github.com/home-assistant/core/blob/5ccc3c17d9d8641edfee3efa266806a3a4025bcd/tests/common.py#L325-L336

Reply:
OK, I've update the test to use this. I've left the other tests alone, as they have both on/off calls going into a single calls array.